src/Microsoft.AspNetCore.Mvc.Core/Internal/ProblemDescriptionFilter.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.ErrorDescription;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ProblemDescriptionFilter : IResultFilter
+    {
+        private readonly IErrorDescriptionFactory _factory;
+
+        public ProblemDescriptionFilter(IErrorDescriptionFactory factory)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+        }
+
+        public void OnResultExecuting(ResultExecutingContext context)
+        {
+            if (!(context.Result is StatusCodeResult statusCodeResult))
+            {
+                return;
+            }
+
+            var statusCode = statusCodeResult.StatusCode;
+            if (statusCode == StatusCodes.Status400BadRequest)
+            {
+                var problem = _factory.CreateErrorDescription(context.ActionDescriptor, new ProblemDescription
+                {
+                    Status = statusCodeResult.StatusCode,
+                    Title = Resources.ProblemDescription_400_Title,
+                });
+
+                context.Result = new BadRequestObjectResult(problem)
```

Comment:
Don't we risk losing valuable data here? If the incoming `context.Result` is a subclass of `StatusCodeResult` that contains extra information that you'd want to surface to the client, then is it possible that this code is discarding that information here?

Reply:
Yup. The check needs to be more stringent.

Reply:
Good point, these probably need to be exact matches. We should be be looking for `NotFoundResult` also